### PR TITLE
Create test_common package

### PR DESCRIPTION
**Pull request by annagrin**
Create test_common package:

- Move `TestSdkLayout` with test-only sdk layout paths to `test_common`.
- Move `SdkAssetGenerator`, logging to `test_common`.
- Move `dwds/test/fixtures/logging.dart` to `test_common`.
- Move sdk asset generator tests to `test_common` tests.
- Update dwds tests to import `test_common`.
- Make `frontend_server_common` use `TestSdkLayout` instead of hard coded layout strings.

Note: this change allows us to enable weak null safety tests for webdev (to be addressed in a separate PR).

Towards: https://github.com/dart-lang/webdev/issues/1878
Closes: https://github.com/dart-lang/webdev/issues/1892

### CODE REVIEW

File: frontend_server_common/lib/src/devfs.dart [hunk 1]
```diff
@@ -4,16 +4,14 @@
 
 // Note: this is a copy from flutter tools, updated to work with dwds tests
 
-import 'package:dwds/asset_reader.dart';
+import 'package:dwds/dwds.dart';
```
Comment (annagrin) [inline frontend_server_common/lib/src/devfs.dart:7]: revert?

File: test_common/analysis_options.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# This file contains the analysis options used for code in the webdev
+# repository.
+
+include: package:lints/recommended.yaml
+
+analyzer:
+  exclude:
+    # Ignore generated files
+    - "lib/data/*"
```
Comment (annagrin) [inline test_common/analysis_options.yaml]: remove irrelevant code

File: test_common/dart_test.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+concurrency: 1
+retry: 3
+
+tags:
```
Comment (annagrin) [inline test_common/dart_test.yaml]: remove

File: test_common/lib/sdk_layout.dart [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+  /// Generate missing assets in the default SDK layout.
+  Future<SdkConfiguration> _create() async {
+    final assetGenerator = SdkAssetGenerator(
+      sdkLayout: sdkLayout,
+      verboseCompiler: _verboseCompiler,
+    );
+
+    if (sdkLayout.soundSummaryPath != sdk.soundSdkSummaryPath) {
+      throw StateError('Invalid asset path ${sdkLayout.soundSummaryPath}');
```
Comment (annagrin) [inline test_common/lib/sdk_layout.dart:217]: remove, we already have tests for this

File: test_common/analysis_options.yaml (hunk 2 above)
Comment (annagrin) [reply test_common/analysis_options.yaml]: done

File: test_common/dart_test.yaml (hunk 3 above)
Comment (annagrin) [reply test_common/dart_test.yaml]: done

File: frontend_server_common/lib/src/devfs.dart (hunk 1 above)
Comment (annagrin) [reply frontend_server_common/lib/src/devfs.dart]: done

File: test_common/lib/sdk_layout.dart (hunk 4 above)
Comment (annagrin) [reply test_common/lib/sdk_layout.dart]: done

File: dwds/lib/dwds.dart [hunk 5]
```diff
@@ -29,4 +29,4 @@ export 'src/services/expression_compiler.dart'
 export 'src/services/expression_compiler_service.dart'
     show ExpressionCompilerService;
 export 'src/utilities/sdk_configuration.dart'
-    show SdkLayout, SdkConfiguration, SdkConfigurationProvider;
+    show SdkConfiguration, SdkConfigurationProvider;
```
Comment (annagrin) [inline dwds/lib/dwds.dart:32]: revert?

File: test_common/lib/sdk_layout.dart [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+SdkConfiguration createConfiguration(TestSdkLayout sdkLayout) =>
+    SdkConfiguration(
+      sdkDirectory: sdkLayout.sdkDirectory,
+      weakSdkSummaryPath: sdkLayout.weakSummaryPath,
+      soundSdkSummaryPath: sdkLayout.soundSummaryPath,
+      compilerWorkerPath: sdkLayout.dartdevcSnapshotPath,
+    );
+
+/// Default sdk configuration.
+final SdkConfiguration defaultSdkConfiguration =
```
Comment (annagrin) [inline test_common/lib/sdk_layout.dart:19]: duplicate, remove

File: test_common/lib/sdk_layout.dart [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+///  - Need to generate SDK summary for weak null safety mode as it
+///    is not provided by the SDK installation.
+///
+/// TODO(annagrin): update to only generating missing sound artifacts
+/// for frontend server after we have no uses of weak null safety.
+class TestSdkConfigurationProvider extends SdkConfigurationProvider {
+  final bool _verboseCompiler;
+  SdkConfiguration? _configuration;
+
+  final _sdk = TestSdkLayout.defaultSdkConfiguration;
```
Comment (annagrin) [inline test_common/lib/sdk_layout.dart:177]: can just inline this

File: dwds/lib/dwds.dart (hunk 5 above)
Comment (annagrin) [reply dwds/lib/dwds.dart]: done

File: test_common/lib/sdk_layout.dart (hunk 7 above)
Comment (annagrin) [reply test_common/lib/sdk_layout.dart]: done

File: dwds/lib/src/utilities/sdk_configuration.dart [hunk 8]
```diff
@@ -37,117 +37,47 @@ abstract class SdkConfigurationProvider {
...
-    required this.weakJsPath,
-    required this.weakJsMapPath,
-    required this.librariesPath,
     required this.dartdevcSnapshotPath,
-    required this.kernelWorkerSnapshotPath,
   });
 }
 
-/// Data class describing the SDK layout.
+/// Sdk configuration.
```
Comment (annagrin) [inline dwds/lib/src/utilities/sdk_configuration.dart:80]: Dart SDK

Comment (annagrin) [reply dwds/lib/src/utilities/sdk_configuration.dart]: done

File: dwds/lib/src/utilities/sdk_configuration.dart [hunk 9]
```diff
@@ -37,117 +37,47 @@ abstract class SdkConfigurationProvider {
...
-  final String sdkJsSoundFileName;
-  final String sdkJsMapSoundFileName;
-  final String sdkSummarySoundFileName;
-  final String sdkSummaryWeakFileName;
-  final String sdkFullDillSoundFileName;
-  final String sdkFullDillWeakFileName;
+  static final defaultSdkDirectory =
+      p.dirname(p.dirname(Platform.resolvedExecutable));
+  static SdkLayout defaultSdkLayout =
+      SdkLayout.createDefault(defaultSdkDirectory);
```
Comment (annagrin) [inline dwds/lib/src/utilities/sdk_configuration.dart:43]: createDefault?

Comment (annagrin) [inline dwds/lib/src/utilities/sdk_configuration.dart:43]: createDefault

File: frontend_server_common/lib/src/resident_runner.dart [hunk 10]
```diff
@@ -36,19 +31,22 @@ class ResidentWebRunner {
...
     bool verbose = false,
   }) {
     generator = ResidentCompiler(
       dartSdkPath,
       projectDirectory: projectDirectory,
       packageConfigFile: packageConfigFile,
       useDebuggerModuleNames: packageUriMapper.useDebuggerModuleNames,
-      platformDill:
-          soundNullSafety ? '$summaryDillSound' : '$summaryDillUnsound',
+      platformDill: soundNullSafety
```
Comment (annagrin) [inline frontend_server_common/lib/src/resident_runner.dart:42]: Uri is needed here instead of file path

Comment (annagrin) [reply frontend_server_common/lib/src/resident_runner.dart]: done

File: dwds/pubspec.yaml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 name: dwds
 # Every time this changes you need to run `dart run build_runner build`.
-version: 18.0.0-dev
+version: 17.0.1-dev
```
[the remark this answers is not in the document]
Comment (annagrin) [reply dwds/pubspec.yaml]: fixed

File: dwds/test/expression_compiler_service_test.dart [hunk 12]
```diff
@@ -9,14 +9,13 @@ import 'dart:convert';
 import 'dart:io';
 
 import 'package:dwds/expression_compiler.dart';
+import 'package:dwds/sdk_configuration.dart';
```
Comment (elliette) [inline dwds/test/expression_compiler_service_test.dart:12]: Should this be `test_sdk_configuration`?

**Review by elliette — APPROVED** [review]
One question but looks good!

File: dwds/test/expression_compiler_service_test.dart (hunk 12 above)
Comment (annagrin) [reply dwds/test/expression_compiler_service_test.dart:12]: This one should not need any extra assets, so it is actually good to test it with the current SDK layout. I'll add a comment.

Comment (annagrin) [reply dwds/test/expression_compiler_service_test.dart:12]: Added!

Comment (annagrin) [reply dwds/test/expression_compiler_service_test.dart:12]: Oh wait it is currently running with weak null safety! Will update (so then the statement that it does not need any extra assets will be true:)

Comment (annagrin) [reply dwds/test/expression_compiler_service_test.dart:12]: done!

**Pull request merged**